### PR TITLE
enable safari extension

### PR DESCRIPTION
**Pull request by ahmedsafadii**
To test the extension:

- Run `Shadowmap Location.xcodeproj`
- Make sure you select your target (Mac, iOS), it support all platforms
Screen Shot 2022-03-02 at 12 58 56 PM
- Click Run
- You would say something like that
Screen Shot 2022-03-02 at 1 02 41 PM

- Click open Safari Extention and enable it by doing the checkmark.

Screen Shot 2022-03-02 at 1 03 58 PM

- Now for Debugging you can from the Developer Menu.
Screen Shot 2022-03-02 at 1 05 00 PM

### CODE REVIEW

**Comment by ahmedsafadii** [comment]
@simonmulser make sure to change the icons, cuz it looks not good atm since I take it from chrome extension.

**Comment by simonmulser** [comment]
> @simonmulser make sure to change the icons, cuz it looks not good atm since I take it from chrome extension.

Did you found somewhere the resolution needed for the image?

**Comment by simonmulser** [comment]
What kind of monster is this PR? :D

**Comment by ahmedsafadii** [comment]
@simonmulser, i didn't do anything just create new project lol, I even remove a lot of useless files which using popup but since you use the action so yeah, the only thing that need to be changes the icons and some of text if you want, for the icon size I already put the icon-{size} so you can replace it, or maybe if you want me to do it for you, just send me your logo with high quality 1024x1024 and I will change it for all icons :)

**Comment by simonmulser** [comment]
shadowmap-logo

Yes, please exchange the pictures! Remember if there is a 2x or 3x you need to use the double or triple resolution. Eg. 20x20@2x you need to use an image which is 40x40 and save it under the filename `iphone-icon-20@2x.png`.

**Comment by simonmulser** [comment]
Steps:
- Open `Shadowmap Location/safari` in XCode
- Select `Shadowmap Location (macOS) > My Mac (Rosetta)`
- Run
- Opens 
image
- Click on `Quit and Open Safari Extension Preferences...`
- Nothing happens


Log:
```
2022-03-02 11:33:22.553868+0100 Shadowmap Location[97188:12705287] Metal API Validation Enabled
```

**Comment by simonmulser** [comment]
Similar if I run it as `Shadowmap Location (macOS) > My Mac (Rosetta)`. There the log looks as follows:
```
2022-03-02 11:35:23.083698+0100 Shadowmap Location[97300:12707703] Metal API Validation Enabled
2022-03-02 11:35:23.094649+0100 Shadowmap Location[97300:12707703] Errors found! Invalidating cache...
```

**Comment by simonmulser** [comment]
image

Please simply use the logo and don't use the grey rectangle in the background.

**Comment by simonmulser** [comment]
@ahmedsafadii I think it makes sense target it only for MacOS. IMHO it doesn't make sense to have this plugin for iOS. What do you think?

**Comment by ahmedsafadii** [comment]
> @ahmedsafadii I think it makes sense target it only for MacOS. IMHO it doesn't make sense to have this plugin for iOS. What do you think?

You can simply ignore it and build it only for macOS, and you need to know that the power of Safari Web Extention, you code one and its release for all types, in case you want it for macOS we can just build for macOS.

**Comment by ahmedsafadii** [comment]
> 

Can you please send me the logo here?

**Comment by ahmedsafadii** [comment]
> Steps:
> 
> * Open `Shadowmap Location/safari` in XCode
> * Select `Shadowmap Location (macOS) > My Mac (Rosetta)`
> * Run
> * Opens
> 
> image
> 
> - Click on `Quit and Open Safari Extension Preferences...` - Nothing happens
> Log:
> 
> ```
> 2022-03-02 11:33:22.553868+0100 Shadowmap Location[97188:12705287] Metal API Validation Enabled
> ```

You need to go to safari browser and then from the preference select extension, you will find it there.

btw we can set a meeting now to run it for you if you want.

**Comment by simonmulser** [comment]
> > Steps:
> > 
> > * Open `Shadowmap Location/safari` in XCode
> > * Select `Shadowmap Location (macOS) > My Mac (Rosetta)`
> > * Run
> > * Opens
> > 
> > image
> > 
> > * Click on `Quit and Open Safari Extension Preferences...` - Nothing happens
> >   Log:
> > 
> > ```
> > 2022-03-02 11:33:22.553868+0100 Shadowmap Location[97188:12705287] Metal API Validation Enabled
> > ```
> 
> You need to go to safari browser and then from the preference select extension, you will find it there.
> 
> btw we can set a meeting now to run it for you if you want.

I saw that. But the button `Quit and Open Safari Extension Preferences...` should work, right? It doesn't work for me. When I click on the button nothing happens.

**Comment by ahmedsafadii** [comment]
> > > Steps:
> > > 
> > > * Open `Shadowmap Location/safari` in XCode
> > > * Select `Shadowmap Location (macOS) > My Mac (Rosetta)`
> > > * Run
> > > * Opens
> > > 
> > > image
> > > 
> > > * Click on `Quit and Open Safari Extension Preferences...` - Nothing happens
> > >   Log:
> > > 
> > > ```
> > > 2022-03-02 11:33:22.553868+0100 Shadowmap Location[97188:12705287] Metal API Validation Enabled
> > > ```
> > 
> > 
> > You need to go to safari browser and then from the preference select extension, you will find it there.
> > btw we can set a meeting now to run it for you if you want.
> 
> I saw that. But the button `Quit and Open Safari Extension Preferences...` should work, right? It doesn't work for me. When I click on the button nothing happens.

If the Extensions preferences window doesn't open automatically, go to the Safari menu and select Preferences. In the preferences window, click on Extensions. The Extensions preference tab opens.

**Comment by simonmulser** [comment]
> > > > Steps:
> > > > 
> > > > * Open `Shadowmap Location/safari` in XCode
> > > > * Select `Shadowmap Location (macOS) > My Mac (Rosetta)`
> > > > * Run
> > > > * Opens
> > > > 
> > > > image
> > > > 
> > > > * Click on `Quit and Open Safari Extension Preferences...` - Nothing happens
> > > >   Log:
> > > > 
> > > > ```
> > > > 2022-03-02 11:33:22.553868+0100 Shadowmap Location[97188:12705287] Metal API Validation Enabled
> > > > ```
> > > 
> > > 
> > > You need to go to safari browser and then from the preference select extension, you will find it there.
> > > btw we can set a meeting now to run it for you if you want.
> > 
> > 
> > I saw that. But the button `Quit and Open Safari Extension Preferences...` should work, right? It doesn't work for me. When I click on the button nothing happens.
> 
> If the Extensions preferences window doesn't open automatically, go to the Safari menu and select Preferences. In the preferences window, click on Extensions. The Extensions preference tab opens.

Yes, I do understand that. Maybe let's discuss the details in a meeting.

Let me know here when you updated the logos!

**Comment by ahmedsafadii** [comment]
@simonmulser Done, i update all icons, i had to do it using sketch :D

File: safari/Shared (Extension)/Resources/_locales/en/messages.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+    "extension_name": {
+        "message": "Shadowmap Location",
+        "description": "Visualize sunlight and shadow anywhere on Earth"
+    },
+    "extension_description": {
+        "message": "The Sun is a prerequisite for life on Earth and Shadowmap is the tool to understand it.",
```
Comment (simonmulser) [inline safari/Shared (Extension)/Resources/_locales/en/messages.json]: Open the current location from Google Maps/Earth in Shadowmap and vice versa.

File: safari/Shared (Extension)/Resources/_locales/en/messages.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
...
+        "description": "Visualize sunlight and shadow anywhere on Earth"
```
Comment (simonmulser) [inline safari/Shared (Extension)/Resources/_locales/en/messages.json]: Shadowmap Location is an extension that allows you to open the current location in Shadowmap with one click in Google Maps. The whole thing also works in the other direction. With a click on the extension icon you can display the current location opened in Google Maps/Earth in Shadowmap.

File: safari/Shared (Extension)/Resources/_locales/en/messages.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+    "extension_name": {
+        "message": "Shadowmap Location",
+        "description": "Visualize sunlight and shadow anywhere on Earth"
```
Comment (simonmulser) [inline safari/Shared (Extension)/Resources/_locales/en/messages.json]: Shadowmap Location allows you to open the current location from Google Maps/Earth in Shadowmap and vice versa.

**Review by simonmulser — CHANGES_REQUESTED** [review]

**Comment by simonmulser** [comment]
image

Can you check if you can simplify the required rights? Currently it says that the plugin can read any information of the website including passwords, browser history. But maybe this is already the least permissive setting.

**Comment by simonmulser** [comment]
> @simonmulser Done, i update all icons, i had to do it using sketch :D

Logos are great!

**Comment by simonmulser** [comment]
Minor and maybe hard to fix: Browser extension icon should be always yellow or yellow when the domain is Google Maps/Earth or Shadowmap. Currently it's only yellow on the tab where it was clicked which is a bit weird.

**Comment by ahmedsafadii** [comment]
> rights

Sad this alert for all types of plugins only on safari :D
Screen Shot 2022-03-02 at 5 33 45 PM

**Comment by simonmulser** [comment]
> > rights
> 
> Sad this alert for all types of plugins only on safari :D Screen Shot 2022-03-02 at 5 33 45 PM

Maybe we can restrict the extension only be active on certain domains (Google Maps/Earth and Shadowmap)? But yeah, if there isn't any other option, let's go for this!

**Comment by ahmedsafadii** [comment]
> > > rights
> > 
> > 
> > Sad this alert for all types of plugins only on safari :D Screen Shot 2022-03-02 at 5 33 45 PM
> 
> Maybe we can restrict the extension only be active on certain domains (Google Maps/Earth and Shadowmap)? But yeah, if there isn't any other option, let's go for this!

https://lapcatsoftware.com/articles/security-safari-extensions.html

```
"permissions": ["activeTab"],
```
This is what we only use we have full permit over correct active tab, but this is normal don't worry.

**Comment by ahmedsafadii** [comment]
> Minor and maybe hard to fix: Browser extension icon should be always yellow or yellow when the domain is Google Maps/Earth or Shadowmap. Currently, it's only yellow on the tab where it was clicked which is a bit weird.

Nah it's something related to memory inactive and active, so the user clicks on it when needs to be loaded in the activeTab.

and maybe because we are in dev mode, but I read that they are inactive the extnetion if it is not needed and the user shall click on it so he understands that it's working.

**Comment by simonmulser** [comment]
> > > > rights
> > > 
> > > 
> > > Sad this alert for all types of plugins only on safari :D Screen Shot 2022-03-02 at 5 33 45 PM
> > 
> > 
> > Maybe we can restrict the extension only be active on certain domains (Google Maps/Earth and Shadowmap)? But yeah, if there isn't any other option, let's go for this!
> 
> https://lapcatsoftware.com/articles/security-safari-extensions.html
> 
> ```
> "permissions": ["activeTab"],
> ```
> 
> This is what we only use we have full permit over correct active tab, but this is normal don't worry.

An option we could try would be using the permission `tabs` (https://developer.chrome.com/docs/extensions/reference/tabs/) instead of `activeTab` and then just read the URL from the tab (https://developer.chrome.com/docs/extensions/reference/tabs/#get-the-current-tab).

But yeah, let's leave it as it is!

**Comment by ahmedsafadii** [comment]
> An option we could try would be using the permission `tabs` (https://developer.chrome.com/docs/extensions/reference/tabs/) instead of `activeTab` and then just read the URL from the tab (https://developer.chrome.com/docs/extensions/reference/tabs/#get-the-current-tab).

I even try that and guess what 

Screen Shot 2022-03-02 at 6 45 46 PM

same message.

**Comment by simonmulser** [comment]
> > An option we could try would be using the permission `tabs` (https://developer.chrome.com/docs/extensions/reference/tabs/) instead of `activeTab` and then just read the URL from the tab (https://developer.chrome.com/docs/extensions/reference/tabs/#get-the-current-tab).
> 
> I even try that and guess what
> 
> Screen Shot 2022-03-02 at 6 45 46 PM
> 
> same message.

This is not the same message Ahmed. This message is way better! It doesn't say anything about sensitive information like passwords and credit cards.

So if we can work with this permissions would be better IMHO. Or am I missing something?

**Comment by ahmedsafadii** [comment]
> > > An option we could try would be using the permission `tabs` (https://developer.chrome.com/docs/extensions/reference/tabs/) instead of `activeTab` and then just read the URL from the tab (https://developer.chrome.com/docs/extensions/reference/tabs/#get-the-current-tab).
> > 
> > 
> > I even try that and guess what
> > Screen Shot 2022-03-02 at 6 45 46 PM
> > same message.
> 
> This is not the same message Ahmed. This message is way better! It doesn't say anything about sensitive information like passwords and credit cards.
> 
> So if we can work with this permissions would be better IMHO. Or am I missing something?

OHHHH yes sorry mb, I had read the above extension message, wait then let me try to get the tab URL from tabs and see.

**Comment by ahmedsafadii** [comment]
Screen Shot 2022-03-02 at 6 59 16 PM

are you ok with this ?

**Comment by simonmulser** [comment]
> Screen Shot 2022-03-02 at 6 59 16 PM
> 
> are you ok with this ?

No, that isn't better. Why does it suddenly show "sensitive information, including ...". We just want to read the URL in the tab and we don't wan't to read the content of the website.

So then let's stick with the initial permission. I think this is even worse.

**Comment by ahmedsafadii** [comment]
@simonmulser zoom?

**Review by simonmulser — APPROVED** [review]

**Pull request merged**